src/set_theory/ordinal_arithmetic.lean

```diff
@@ -894,38 +908,174 @@ match o, o.out, o.out_eq, f :
   ⟨λ H i h, by simpa only [typein_enum] using H (enum r i h), λ H b, H _ _⟩
 end
 
+theorem le_bsup {o} (f : Π a < o, ordinal) (i h) : f i h ≤ bsup o f :=
+bsup_le.1 (le_refl _) _ _
+
+theorem lt_bsup {o} (f : Π a < o, ordinal) {a} : a < bsup o f ↔ ∃ i hi, a < f i hi :=
+by simpa only [not_forall, not_le] using not_congr (@bsup_le _ f a)
+
 theorem bsup_type (r : α → α → Prop) [is_well_order α r] (f) :
   bsup (type r) f = sup (λ a, f (typein r a) (typein_lt_type _ _)) :=
 eq_of_forall_ge_iff $ λ o,
 by rw [bsup_le, sup_le]; exact
   ⟨λ H b, H _ _, λ H i h, by simpa only [typein_enum] using H (enum r i h)⟩
 
-theorem le_bsup {o} (f : Π a < o, ordinal) (i h) : f i h ≤ bsup o f :=
-bsup_le.1 (le_refl _) _ _
+theorem sup_eq {ι} {f : ι → ordinal} :
+  sup f = bsup (type well_ordering_rel) (λ a ha, f (enum well_ordering_rel a ha)) :=
+by simp [bsup_type]
+
+theorem is_normal.bsup {f} (H : is_normal f) {o} :
+  ∀ (g : Π a < o, ordinal) (h : o ≠ 0), f (bsup o g) = bsup o (λ a h, f (g a h)) :=
+induction_on o $ λ α r _ g h,
+by resetI; rw [bsup_type,
+     H.sup (type_ne_zero_iff_nonempty.1 h), bsup_type]
 
-theorem lt_bsup {o : ordinal} {f : Π a < o, ordinal}
-  (hf : ∀{a a'} (ha : a < o) (ha' : a' < o), a < a' → f a ha < f a' ha')
+-- A result that shows up twice for some reason.
+private lemma lt_bsup_of_ne_bsup {o : ordinal} {f : Π a < o, ordinal} :
+  (∀ i h, f i h ≠ o.bsup f) → ∀ i h, f i h < o.bsup f :=
+λ hf _ _, lt_of_le_of_ne (le_bsup _ _ _) (hf _ _)
+
+theorem bsup_not_succ_of_ne_bsup {o} {f : Π a < o, ordinal}
+  (hf : ∀ (i : ordinal) (h : i < o), f i h ≠ o.bsup f) (a) :
+  a < bsup o f → succ a < bsup o f :=
+begin
+  replace hf := lt_bsup_of_ne_bsup hf,
+  intro hao,
+  by_contra' hoa,
+  have hao' := le_antisymm (succ_le.2 hao) hoa,
+  rw ←hao' at hf,
+  rw le_antisymm (le_of_lt hao) (bsup_le.2 (λ i h, lt_succ.1 (hf i h))) at hao',
+  exact succ_ne_self _ hao',
+end
+
+theorem lt_bsup_of_limit {o : ordinal} {f : Π a < o, ordinal}
```

Comment:
Is there `sup` version of this result?

Reply:
This result depends both on the fact the domain is ordered, and on the fact that it's limited by a limit ordinal, so I don't think there's anything natural. Since `lt_bsup_of_limit` is currently not even used, there's not really any concrete usage to compare with.

Reply:
Maybe I could prove something similar about a strictly monotone function `f : ι → ordinal` where `ι` is a `linear_order` and a `succ_order` such that `a < succ a`? But again, this seems way too contrived to ever be useful in practice.

Reply:
Fair point!